R/capture.R

```diff
@@ -1,51 +1,198 @@
+    hidden_functions <- setdiff(all_functions, exported_function)
+
+    lapply(hidden_functions, function(func) {
+        decorate_function(func, env = pkg_namespace)
+    })
+}
+
+#' @title Undecorate all functions.
+#'
+#' @description Undecorate all functions.
+#' Functions decorated through `decorate_function_val()` are not undecorated.
+#' @export
+#'
+undecorate_all <- function() {
+    lapply(cache$decorated_functions, function(decorated_fun) {
+        if (decorated_fun$type == "value") {
+            ; # do nothing
+        } else if (decorated_fun$type == "bound") {
+            env <- decorated_fun$env
+            name <- decorated_fun$name
+            env[[name]] <- decorated_fun$original_func
+        } else if (decorated_fun$type == "exported") {
+            package <- decorated_fun$package
+            name <- decorated_fun$name
+            overwrite_export(name, decorated_fun$original_func, package)
+        }
+    })
+    cache$decorated_functions <- list()
 }
 
-isDecoratedFun <- function(fn) {
+#' @title Tells you whether the function is a result of genthat's function decoration.
+#'
+#' @param fn function value
+#' @export
+#'
+is_decorated <- function(fn) {
     if (typeof(fn) != "closure") {
         FALSE
     } else {
         b <- body(fn)
         if (length(b) != 3) {
             FALSE
         } else {
-            tracer.expr <- b[[2]]
-            isTRUE(attr(tracer.expr, 'isGenthatTracerExpression'))
+            tracer_expr <- b[[2]]
+            isTRUE(attr(tracer_expr, 'isGenthatTracerExpression'))
         }
     }
 }
 
-#' @title Decorates function to capture calls and return values
+##########################
+# NON-EXPORTED FUNCTIONS #
+##########################
+
+#' @title Returns traced version of original function.
 #'
 #' @description More sane replacement of the base::trace function.
 #' @param func function value
-#' @param func_name function name
-#' @export
+#' @param func_label name that will be recorded in traces
+#' @param enter_function name that will be recorded in traces
+#' @param exit_function name that will be recorded in traces
 #'
-decorate_function_val <- function(func, func_name, ..., enter_function, exit_function) {
-    if (isDecoratedFun(func)) return(func) #stop("Trying to decorate already decorated function!")
-
-    tracer.expr <- substitute({
-        call_id <- call_id_counter$value
-        assign("value", call_id + 1, call_id_counter)
+decorate_function_val__ <- function(func, func_label, enter_function, exit_function) {
+    if (is_decorated(func)) return(func) #stop("Trying to decorate already decorated function!")
 
-        forceArgs <- FALSE
+    tracer_expr <- substitute({
+        call_id <- genthat:::gen_cid()
+        
+        forceArgs <- TRUE
 
         if (forceArgs) {
             e <- environment()
-            arg_names <- names(as.list(match.call())[-1])
+            arg_names <- names(as.list(match.call())[-1]) # TODO maybe call formals ?
```

Comment:
`formals()` on the other hand doesn't return names of parameters passed as `...`. 
But combining `formals()` and `names(substitute(alist(...)))` could work.